woocommerce/woocommerce
# Fix: Multiple issues related to MySQL having ANSI_QUOTES enabled in Product Filters

### Changes proposed in this Pull Request:





Closes #51216 

This PR addresses multiple issues that arise when MySQL has the `ANSI_QUOTES` mode enabled. The `ANSI_QUOTES` mode treats double quotes as identifier quotes instead of string quotes, which causes problems with our SQL queries.

The solution involves modifying our SQL queries to use backticks for column and table names instead of double quotes. This ensures that the queries work correctly regardless of whether `ANSI_QUOTES` is enabled or not.

Here are some of the errors which were showing in `error.log` file:
```
[17-Oct-2024 06:44:14 UTC] WordPress database error Unknown column 'pa_color' in 'where clause' for query 
			SELECT COUNT( DISTINCT posts.ID ) as term_count, terms.term_id as term_count_id
			FROM wp_posts AS posts
			INNER JOIN wp_term_relationships AS term_relationships ON posts.ID = term_relationships.object_id
			INNER JOIN wp_term_taxonomy AS term_taxonomy USING( term_taxonomy_id )
			INNER JOIN wp_terms AS terms USING( term_id )
			WHERE posts.ID IN ( SELECT   wp_posts.ID
					 FROM wp_posts 
					 WHERE 1=1  AND wp_posts.post_type = 'product' AND ((wp_posts.post_status = 'publish'))
					 
					 ORDER BY wp_posts.post_date DESC, wp_posts.ID DESC
					  )
			AND term_taxonomy.taxonomy IN ("pa_color")
			GROUP BY terms.term_id
```

```
[17-Oct-2024 07:59:20 UTC] WordPress database error Unknown column 'outofstock' in 'where clause' for query 
			SELECT min( min_price ) as min_price, MAX( max_price ) as max_price
			FROM wp_wc_product_meta_lookup
			WHERE product_id IN ( SELECT   wp_posts.ID
					 FROM wp_posts  LEFT JOIN wp_wc_product_meta_lookup wc_product_meta_lookup ON wp_posts.ID = wc_product_meta_lookup.product_id 
					 WHERE 1=1  AND wp_posts.post_type = 'product' AND ((wp_posts.post_status = 'publish')) AND wc_product_meta_lookup.stock_status IN ("outofstock")
					 
					 ORDER BY wp_posts.post_date DESC, wp_posts.ID DESC
					  )
		 made by require('wp-blog-header.php'), wp, WP->main, WP->parse_request, do_action_ref_array('parse_request'), WP_Hook->do_action, WP_Hook->apply_filters, rest_api_loaded, WP_REST_Server->serve_request, WP_REST_Server->dispatch, WP_REST_Server->respond_to_request, Automattic\WooCommerce\StoreApi\Routes\V1\AbstractRoute->get_response, Automattic\WooCommerce\StoreApi\Routes\V1\AbstractRoute->get_response_by_request_method, Automattic\WooCommerce\StoreApi\Routes\V1\ProductCollectionData->get_route_response, Automattic\WooCommerce\StoreApi\Utilities\ProductQueryFilters->get_filtered_price, QM_DB->query
```




### How to test the changes in this Pull Request:



**Testing "Filter by stock" block on Frontend**

1. Edit `my.cnf` to add `sql_mode=ANSI_QUOTES`
2. On "Product Catalog" template, add "Filter by stock" block
3. On Frontend shop page, verify that:
	- "In stock" checkbox works as expected
	- "Out of stock" checkbox works as expected
	- *Sidenote:* Make sure you have some products in stock and out of stock to test this.
4. On trunk branch (For developers only)
	- Verify that reported issue is reproducible. Follow the steps from reported issue: #51216
	- On trunk, it should show an error message similar to the one displayed in the screenshot below:
![image](https://github.com/user-attachments/assets/a02e2c42-c56f-4919-9ffb-767651afe128)


**Testing "Product Filters" pattern**
1. On "Product Catalog" template, add "Product Filters" pattern
2. Verify that there are no errors on `error.log` file related to Database which looks like this:
```
[17-Oct-2024 06:44:14 UTC] WordPress database error Unknown column 'pa_color' in 'where clause' for query 
			SELECT COUNT( DISTINCT posts.ID ) as term_count, terms.term_id as term_count_id
			FROM wp_posts AS posts
```
3. Publish the changes
4. On Frontend, test product filters like "Filter by Stock", "Filter by Attributes"
	- Verify that all of them work as expected
	- There are no errors on `error.log` file related to Database



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

Fix: Multiple issues related to MySQL having `ANSI_QUOTES` enabled in Product Filters

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
